### ISSUE TITLE
[bug] HMR only works on base URL

### ISSUE THREAD

**Issue by mensch** [roots/bud#1070]
### Agreement

- [X] This is not a duplicate of an existing issue
- [X] I have read the [guidelines for Contributing to Roots Projects](https://github.com/roots/.github/blob/master/CONTRIBUTING.md)
- [X] This is not a personal support request that should be posted on the [Roots Discourse](https://discourse.roots.io/) community

### Expected Behavior

On http://localhost:3000/my-subpage, HMR should update styles and load resources through the proxy URL and update on changes.

### Actual Behavior

HMR runs, but there are several messages about CORS violations (related to fonts and images). When updating styles a `*.hot-update.js` is correctly generated, but it can't be found by the server: `Cannot find update (Full reload needed)`. The server tries to load the resource from the URL with the appended path.

### Steps To Reproduce

1. Run `yarn dev`
2. Navigate to a subpage, e.g. http://localhost:3000/my-subpage
3. Make a change in SCSS
4. Build fails, tries to load update from _http://localhost:3000/my-subpage/*.hot-update.js_ instead of _http://localhost:3000/*.hot-update.js_

### version

@roots/bud/5.2.0 darwin-x64 node-v17.4.0

### What package manager are you using?

yarn berry

### version

3.1.1

### Are you using pnpm?

- [ ] yes
- [X] no

### Configuration

```shell
const SVGSpritemapPlugin = require('svg-spritemap-webpack-plugin');

/**
 * @typedef {import('@roots/bud').Bud} bud
 *
 * @param {bud} app
 */
module.exports = (app) =>
  app
    /**
     * Application entrypoints
     *
     * Paths are relative to your resources directory
     */
    .entry({
      app: ['scripts/app.js', 'styles/app.scss'],
      editor: ['scripts/editor.js', 'styles/editor.scss'],
    })

    /**
     * These files should be processed as part of the build
     * even if they are not explicitly imported in application assets.
     */
    .assets([app.path('src', 'images')])

    /**
     * These files will trigger a full page reload
     * when modified.
     */
    .watch(['resources/views/**/*.blade.php', 'app/View/**/*.php'])

    /**
     * Target URL to be proxied by the dev server.
     *
     * This is your local dev server.
     */
    .proxy('http://current-dev-domain.test')

    /**
     * Generate SVG sprite based on available icons
     *
     */
    .use(
      new SVGSpritemapPlugin('resources/images/icons/*.svg', {
        output: {
          filename: 'icons.svg',
          chunk: {keep: true},
          svgo: false,
        },
        sprite: {
          prefix: false,
        },
      }),
    );
```

**Comment by mensch** [comment]
I've pulled the changes from [fix-bud-hmr](https://github.com/roots/acorn/pull/178) and they work with one change. My `hmr.json` has `localhost:3000` with a trailing slash, which makes HMR bug out because of the double slashes in the resulting URL. If I remove the slash in hmr.json style and script updates work. Loading external resources (like fonts) still seems to be broken, it tries to load those files at a subpath instead of the root URL.

I'm not that well-versed in the whole architecture of Bud and Acorn, but maybe `getBudDevUri` in `/Acorn/Assets/Middleware/RootsBudMiddleware.php` should remove any trailing slashes before passing the URL?

**Comment by kellymears** [comment]
> I've pulled the changes from [fix-bud-hmr](https://github.com/roots/acorn/pull/178) and they work with one change. My `hmr.json` has `localhost:3000` with a trailing slash, which makes HMR bug out because of the double slashes in the resulting URL. If I remove the slash in hmr.json style and script updates work. Loading external resources (like fonts) still seems to be broken, it tries to load those files at a subpath instead of the root URL.
> 
> I'm not that well-versed in the whole architecture of Bud and Acorn, but maybe `getBudDevUri` in `/Acorn/Assets/Middleware/RootsBudMiddleware.php` should remove any trailing slashes before passing the URL?

I think `fix-bud-hmr` will break until I submit a fix for the publicPath set on bud's end. Somewhere on boot in `@roots/sage`:

https://github.com/roots/bud/blob/a10645a8d3a3ade4f082e89ea0fe601628c21f71/sources/%40roots/sage/src/sage.preset.ts#L39-L92

**Comment by kellymears** [comment]
this is fully fixed as of 5.3.2 w/ acorn @2.

**Comment by mensch** [comment]
5.7.3 seems to have introduced a regression. When I run `yarn dev` all resources (webfonts, etc.) are properly loaded on the main URL, but they're redirected when visiting a subpage. I've also upgraded to the latest version of acorn (2.0.5), but that doesn't seem to resolve things.

**Comment by retlehs** [comment]
@mensch confirmed on our end, thanks for message!

**Comment by retlehs** [comment]
temporary workaround via @kellymears is to define the `publicPath` for the entrypoints:

```js
    /**
     * Application entrypoints
     */
    .entry({
      app: {
        import: ["{scripts,styles}/app.*"],
        publicPath: "/app/themes/sage/public/",
      },
      editor: {
        import: ["{scripts,styles}/editor.*"],
        publicPath: "/app/themes/sage/public/",
      },
    })
```

**Comment by retlehs** [comment]
fixed in 5.7.4

**Comment by mensch** [comment]
I can confirm this is indeed working. Loading images inlined in the HTML still seems to be broken however. The resource is requested over `http://localhost:3000/wp-content/uploads` and returns a success status (200) but the response is empty.. The same image is successfully loaded over `http://my-website.test/wp-content/uploads`.

**Comment by retlehs** [comment]
@mensch should be fixed once we get #1319 into a release soon

**Comment by mensch** [comment]
I see #1319 has been merged. I've updated to 5.7.7 and saw that Sage has updated its development URL to: `http://0.0.0.0:3000`.

This indeed makes images work, but it leads to CORS errors (at least in Chrome) when doing XMLHttpRequests: _Access to XMLHttpRequest at 'http://testdomain.test/wp-json/' from origin 'http://0.0.0.0:3000' has been blocked by CORS policy: The request client is not a secure context and the resource is in more-private address space `local`._

**Comment by retlehs** [comment]
@mensch can you use browsersync for now?


1. install browsersync: `yarn add browser-sync browser-sync-webpack-plugin --dev`
2. modify `bud.config.js`:

```javascript
# @ bud.config.js
const bs = require('browser-sync-webpack-plugin')

...
    .use(new bs({proxy: 'http://example.test'}))
```

<details>
 <summary>full config example</summary>

```javascript
const bs = require('browser-sync-webpack-plugin')

module.exports = async (app) => {
  app
    .entry({
      app: ['@scripts/app', '@styles/app'],
      editor: ['@scripts/editor', '@styles/editor'],
    })
    .assets('images')
    .watch('resources/views/**/*', 'app/**/*')
    .use(new bs({proxy: 'http://example.test'}))
};
```

</details>

**Comment by mensch** [comment]
Using BrowserSync would work, but unfortunately I can't get it to run. When I use the configuration provided it says that the address is already in use. When I change the port, the script runs, but none of the relative paths work.

Substituting `0.0.0.0:3000` with `localhost:3000` is a workaround for now, as I can test my XMLHttpRequests without the images loading.
